NuRadioReco/modules/phasedarray/phasedArrayBase.py

```diff
@@ -83,32 +84,71 @@ def calculate_time_delays(
         beam_rolls: array of dicts of keys=antenna and content=delay
         """
 
+        if station.get_id() in self.buffered_delays:
+            return self.buffered_delays[station.get_id()]
+
         if triggered_channels is None:
             triggered_channels = [channel.get_id() for channel in station.iter_trigger_channels()]
 
-        time_step = 1. / sampling_frequency
-
         ant_z = self._get_antenna_positions(station, det, triggered_channels, 2)
-
         self.check_vertical_string(station, det, triggered_channels)
         ref_z = np.max(ant_z)
-
-        # Need to add in delay for trigger delay
-        cable_delays = np.array([det.get_cable_delay(station.get_id(), channel_id) for channel_id in triggered_channels])
+        cable_delays = np.array([det.get_cable_delay(station.get_id(), channel_id, trigger=True) for channel_id in triggered_channels])
+        group_delays = self.get_response_group_delays(station, det, triggered_channels, fmin=0.15, fmax=0.2)
 
         beam_rolls = []
-        for angle in phasing_angles:
-
-            delays = -(ant_z - ref_z) / cspeed * ref_index * np.sin(angle) - cable_delays
-
-            delays -= np.max(delays)
-            roll = np.array(np.round(delays / time_step)).astype(int)
 
+        for angle in phasing_angles:
+            delays = (ant_z - ref_z) / cspeed * ref_index * np.sin(angle) - cable_delays - group_delays
+            delays -= np.min(delays)
+            roll = np.array(np.round(delays * sampling_frequency)).astype(int)
             subbeam_rolls = dict(zip(triggered_channels, roll))
             beam_rolls.append(subbeam_rolls)
 
+        self.buffered_delays[station.get_id()] = beam_rolls
+
         return beam_rolls
 
+    def get_response_group_delays(self, station, det, triggered_channels, fmin=0.150, fmax=0.200):
```

Comment:
as mentioned above, there is a class function defined for the response class. I think I would rather improve this function instead of defining a new one.